text/0000-const-trait-impls.md

```diff
@@ -0,0 +1,1006 @@
+
+The disadvantage of this is that by default, it creates stricter bounds than desired.
+
+```rust
+const fn foo<T: Foo>() {
+    T::ASSOC_CONST
+}
+```
+
+compiles today, and allows all types that implement `Foo`, irrespective of the constness of the impl.
+With the opt-out scheme that would still compile, but suddenly require callers to provide a const impl.
+
+The safe default (and the one folks are used to for a few years now on stable), is that trait bounds just work, you just
+can't call methods on them.
+This is both useful in
+
+* nudging function authors to using the minimal necessary bounds to get their function
+body to compile and thus requiring as little as possible from their callers,
+* ensuring our implementation is correct by default.
+
+The implementation correctness argument is partially due to our history with `?const` (see https://github.com/rust-lang/rust/issues/83452 for where we got it wrong and thus decided to stop using opt-out), and partially with our history with `?` bounds not being great either (https://github.com/rust-lang/rust/issues/135229, https://github.com/rust-lang/rust/pull/132209). An opt-in is much easier to make sound and keep sound.
+
+To get more capabilities, you add more syntax. Thus the opt-out approach was not taken.
+
+## Per-method constness instead of per-trait
+
+We could require trait authors to declare which methods can be const:
+
+```rust
+trait Default {
+    const fn default() -> Self;
+}
+```
+
+This has two major advantages:
+
+* you can now have const and non-const methods in your trait without requiring an opt-out
+* you can add new methods with default bodies and don't have to worry about new kinds of breaking changes
+
+The specific syntax given here may be confusing though, as it looks like the function is always const, but
+implementations can use non-const impls and thus make the impl not usable for `T: ~const Trait` bounds.
+
+Though this means that changing a non-const fn in the trait to a const fn is a breaking change, as the user may
+have that previous-non-const fn as a non-const fn in the impl, causing the entire impl now to not be usable for
+`T: ~const Trait` anymore.
+
+See also: out of scope RTN notation in [Unresolved questions](#unresolved-questions)
+
+## Per-method and per-trait constness together:
+
+To get the advantages of the per-method constness alternative above, while avoiding the new kind of breaking change, we can require per-method and per-trait constness:
+
+A mixed version of the above could be
+
+```rust
+const trait Foo {
+    const fn foo();
+    fn bar();
+}
+```
+
+where you still need to annotate the trait, but also annotate the const methods.
+
+But it makes it much harder/more confusing to add
+
+```rust
+trait Tr {
+    const C: u8 = Self::f();
+    const fn f() -> u8;
+}
+```
+
+later, where even non-const traits can have const methods, that all impls must implement as a const fn.
```

Comment:
Another possible syntax is:

```rust
const trait Tr {
    // Must be `const` in all impls
    const fn foo();
    // Must be `const` in `const` impls
    ~const fn bar();
    // Never needs to be `const`
    fn baz();
}
```

This is analogous to the syntax for trait bounds.

A further possibility is to not require the `const` keyword in front of the trait for `impl`s, but instead infer based on the methods:

```rust
// This impl is `const`, because `fn bar()` is
impl trait Tr for () {
    const fn foo() {}
    const fn bar() {}
    fn baz() {}
}

// This impl is not `const`, because `fn bar()` is not
impl trait Tr for u32 {
    const fn foo() {}
    fn bar() {}
    fn baz() {}
}
```

Reply:
It seems very suprising to have such aggregate behaviour where individual method annotations change sth about the entire impl.

We have that for fields, but there's no precedent for impls and it seems a footgun

Reply:
Depends how we interpret what `const Trait` means; we could mean it as "the methods can be called in a const context". This would solve the `~const Copy` issues because that would make `T: const Copy` and `T: Copy` equivalent.

Reply:
In my proposal above, `const Trait` bounds are basically sugar for the RTN-like syntax discussed at e.g. https://github.com/rust-lang/rfcs/pull/3762#discussion_r1922775028. In my example above, bounding by `const Tr` is equivalent to defining a trait alias `trait ConstTr = Tr where Self::bar: const;`, and then bounding by `ConstTr`.

Reply:
Arguably “it’s just an alias” also makes the feature easier to teach.

> It also imposes the limitation that a `const` trait must contain at least one `~const` function or bound.

Alternatively, we could allow it with a warning, so that a `const Trait` bound for a `const trait` with no `~const` in its definition is equivalent to the bare `Trait` bound.

Reply:
By first syntax it seems that there is no need to mark a trait as `const`, since `fn` methods' default bodies are unnecessary to being const.

Reply:
One question with my proposal is how to handle default method bodies whose constness depends on that of other methods, like all the `Iterator` helpers. Here, I imagine that `~const` bounds on a method declaration would be considered to apply to the default impl. So, for example, `count()` from `Iterator` would be declared as `~const fn count(self) -> usize where Self: ~const Iterator { /* ... */ }`. (This requires [coinduction](https://rustc-dev-guide.rust-lang.org/solve/coinduction.html).)

Going further, Perhaps `where ~const` could be added as a shorthand for `where Self: ~const TheTraitBeingDefined`. So our `count()` declaration could be written `~const fn count(self) -> usize where ~const { /* ... */ }`.

Reply:
Most of my suggestions here were adopted. This thread should be resolved in favor of https://github.com/rust-lang/rfcs/pull/3762#discussion_r1985279156 and https://github.com/rust-lang/rfcs/pull/3762#discussion_r1985280055.